### ISSUE TITLE
Dynamic DNS updates cannot use Multi_WAN interface

### ISSUE THREAD

**Issue by bendiy** [opnsense/plugins#1270]
Versions:
* OPNsense 18.7.7-amd64
* FreeBSD 11.1-RELEASE-p15
* OpenSSL 1.0.2p 14 Aug 2018

I'm not sure if this is a `core` issue because of Multi_WAN or a Dynamic DNS issues in `plugins`.

I have setup OpnSense with Multi_WAN fail over, which is working great. I have two WANs setup with the first as the default gateway and the second as a tier 2 fail over gateway. Disconnecting `WAN1` (`5.6.7.8`) fails over to `WAN2` (`1.2.3.4`).

I'm now trying to configure Dynamic DNS updates of a subdomain, let's call it `dyn.example.com`. Initially, I added two Dynamic DNS entries under `Services > Dynamic DNS`. The first entry was for `WAN1` and the second for `WAN2`. I successfully testing the update using `Save and Force Update`, which updates the public DNS entry for `dyn.example.com` to the public IP of `WAN1` or `WAN2`.

Now I'm trying to get Dynamic DNS updates to work seamlessly when OpnSense fails over from `WAN1` to `WAN2`. When I disconnect `WAN1`, everything fails over and the log shows this:

```
Dec 12 10:00:42 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): No change in my IP address and/or 25 days has not passed. Not updating dynamic DNS entry.
Dec 12 10:00:42 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): Current WAN IP: 1.2.3.4 Cached IP: 1.2.3.4
Dec 12 10:00:42 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): 1.2.3.4 extracted
Dec 12 10:00:42 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): running get_failover_interface for opt2. found em3
Dec 12 10:00:42 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): 1.2.3.4 extracted
Dec 12 10:00:42 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS: updatedns() starting
```

Fair enough, the public IP of `WAN2` is still `1.2.3.4`, which is cached by OpnSense, so it isn't going to actually make the Dynamic DNS update request. Not the desired result, but it makes _sense_ what's happening.

Next idea is that maybe I should have only one Dynamic DNS entry under `Services > Dynamic DNS` and use the `Multi_WAN` interface instead of `WAN1` and `WAN2`. I set this up and now when I test fail over, I get this in the logs:

```
Dec 12 10:06:33 | opnsense: /services_dyndns_edit.php: Curl error occurred: Couldn't bind to ''
Dec 12 10:06:33 | opnsense: /services_dyndns_edit.php: Dynamic DNS (dyn.example.com): Current Service: googledomains
Dec 12 10:06:33 | opnsense: /services_dyndns_edit.php: Dynamic DNS (dyn.example.com): _checkStatus() starting.
Dec 12 10:06:33 | opnsense: /services_dyndns_edit.php: Dynamic DNS (dyn.example.com via MyDNSProviderHere): _update() starting.
Dec 12 10:06:33 | opnsense: /services_dyndns_edit.php: Dynamic DNS (dyn.example.com): running get_failover_interface for Multi_WAN. found opt2_vip5
Dec 12 10:06:33 | opnsense: /services_dyndns_edit.php: Dynamic DNS (dyn.example.com): 1.2.3.4 extracted
Dec 12 10:06:32 | opnsense: /services_dyndns_edit.php: Dynamic DNS: updatedns() starting
```

`Curl error occurred: Couldn't bind to ''`, it looks like `Curl` cannot use the `Multi_WAN` interface for the HTTP request. The code sets the `Curl` interface like this:

https://github.com/opnsense/plugins/blob/61ff5956d5f7a79c6609d5c34e7d74e0c15b3f79/dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc#L266-L270
```php
$this->_dnsRequestIf = get_failover_interface($dnsRequestIf);
if ($this->_dnsVerboseLog) {
     log_error("Dynamic DNS ({$this->_dnsHost}): running get_failover_interface for {$dnsRequestIf}. found {$this->_dnsRequestIf}");
}
$this->_dnsRequestIfIP = get_interface_ip($dnsRequestIf);
```
https://github.com/opnsense/plugins/blob/61ff5956d5f7a79c6609d5c34e7d74e0c15b3f79/dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc#L361
```php
curl_setopt($ch, CURLOPT_INTERFACE, $this->_dnsRequestIfIP);
```

Is `Multi_WAN` not a valid interface for `Curl`?

Playing around some more, I tried setting `Service type = Custom`  for Dynamic DNS, which gives you two options for the interface, `Interface to monitor` and `Interface to send update from`. I was able to get a Multi_WAN fail back to normal (`WAN2` to `WAN1`) DNS update to work by adding two `Custom` entries under `Services > Dynamic DNS`:
1. `Interface to monitor = Multi_WAN` and `Interface to send update from = WAN1`.
2. `Interface to monitor = Multi_WAN` and `Interface to send update from = WAN2`.

```
Dec 12 11:42:01 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS: (Success) IP Address Updated Successfully!
Dec 12 11:42:01 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS: updating cache file /var/cache/dyndns_Multi_WAN_dyn.example.com_2.cache: 5.6.7.8
Dec 12 11:42:01 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): 5.6.7.8 extracted
Dec 12 11:41:58 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): Current Service: custom
Dec 12 11:41:58 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): _checkStatus() starting.
Dec 12 11:41:58 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com via Custom): _update() starting.
Dec 12 11:41:58 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): running get_failover_interface for opt2. found em3
Dec 12 11:41:58 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): 5.6.7.8 extracted
Dec 12 11:41:58 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS: updatedns() starting
```

But when I test fail over again (`WAN1` to `WAN2`), I get a timeout error from `Curl`. I think this may be due to `WAN2` not being fully "failed over to" yet when the Dynamic DNS update runs:

```
Dec 12 11:01:49 | opnsense: /usr/local/etc/rc.dyndns: Curl error occurred: Connection timed out after 15010 milliseconds
Dec 12 11:01:49 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): Current Service: custom
Dec 12 11:01:49 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): _checkStatus() starting.
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com via Custom): _update() starting.
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): running get_failover_interface for opt2. found em3
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS (dyn.example.com): 1.2.3.4 extracted
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: GATEWAYS: Group Multi_WAN did not have any gateways up on tier 1!
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: MONITOR: WAN1 is down, removing from routing group Multi_WAN
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: GATEWAYS: Group Multi_WAN did not have any gateways up on tier 1!
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: MONITOR: WAN1 is down, removing from routing group Multi_WAN
Dec 12 11:01:34 | opnsense: /usr/local/etc/rc.dyndns: Dynamic DNS: updatedns() starting
```

Note that both show `running get_failover_interface for opt2. found em3`.

Failing over from `WAN1` to `WAN2` seems to be slow (10 to 60 seconds vs. immediately) to establish the correct routing in the firewall compared to going from `WAN2` to `WAN1` when `WAN1` is the default.

Some ideas on what might be needed:
1. Add both interface options to all Dynamic DNS services, `Interface to monitor` and `Interface to send update from` so `Custom` does not need to be used for this.
2. Add an ability to disable the cached IP check and just update the DNS anyways. Is there any harm in always updating it? How often does this update run when not in a fail over or IP changed situation?
3. Add a retry on error interval to the Dynamic DNS service to give `WAN2` some time to come up completely during a fail over. If I give `WAN2` a minute to come up and then run the `Save and Force Update`, it works fine. I'm not sure if the delay in `WAN2` coming up during a fail over (which causes the `Curl` timeout) is typical of any Multi_WAN config or unique to my config. I currently cannot do something like `ping google.com` on `WAN2` when `WAN1` is up and `WAN1` is the primary gateway. I can only ping the `WAN2`'s gateway `Monitor IP`, `8.8.8.8`, that I have a firewall rule for. Once `WAN1` is down and `WAN2` is fully up, `ping google.com` works from `WAN2`. Do I need an additional firewall rule or route to always allow the Dynamic DNS request out on `WAN2`?

What can I do to get Dynamic DNS updates to work when `WAN1` fails over to  `WAN2`?

Thanks!

**Comment by alexanderharm** [comment]
Does someone have a working example for this? We have a dual WAN setup, one via DHCP (receives a public ip), one static but with a private network ip since it is behind another router. We defined a gateway group but DynDNS won't update if group is selected as monitored gateway.

**Comment by AdSchellevis** [comment]
@alexanderharm dyndns with gateway groups is broken https://github.com/opnsense/plugins/issues/1416, it's a very old plugin, could definitely use a rewrite hooking into something more standard if possible, but needs a contributor.

**Comment by AdSchellevis** [comment]
This issue has been automatically timed-out (after 180 days of inactivity).

For more information about the policies for this repository,
please read https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md for further details.

If someone wants to step up and work on this issue,
just let us know, so we can reopen the issue and assign an owner to it.

**Comment by sepahewe** [comment]
Fixed in 19.7.9

**Comment by bramevo** [comment]
> Does someone have a working example for this? We have a dual WAN setup, one via DHCP (receives a public ip), one static but with a private network ip since it is behind another router. We defined a gateway group but DynDNS won't update if group is selected as monitored gateway.

Same issue here on current version. For some reason, it's not updating the IP of the fallback WAN unless I disable and enable the DynDNS manually. Although the service sees that the old IP is no longer valid, it won't update it.

**Comment by mayerthomas** [comment]
If fixed in 19.7.9 I also would be interested in a working example.
Can somebody help?
